stout/range.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include <algorithm>
+
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+namespace detail {
+
+////////////////////////////////////////////////////////////////////////
+
+struct _Range {
+  template <typename K_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, int from, int to)
+      : k_(std::move(k)),
+        to_(to),
+        from_(std::min(from, to)) {}
+
+    void Start() {
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    void Ended() {
+      k_.Ended();
+    }
+
+    void Next() override {
+      if (from_ == to_) {
+        k_.Ended();
+      } else {
+        k_.Body(from_++);
+      }
+    }
+
+    void Done() override {
+      k_.Ended();
+    }
+
+    K_ k_;
+    int to_;
+    int from_;
+  };
+
+  template <typename T>
+  struct Composable {
+    template <typename>
+    using ValueFrom = T;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), from_, to_};
+    }
+
+    int from_;
+    int to_;
+  };
+};
+
+
+////////////////////////////////////////////////////////////////////////
+
+} // namespace detail
+
+////////////////////////////////////////////////////////////////////////
+
+auto Range(int from, int to) {
+  return detail::_Range::Composable<int>{from, to};
+}
+
```

Comment:
Shall we also add a:
```c++
auto Range(int to) {
```
... Which assumes that `from` is `0`? That matches behavior on e.g. Python's `for i in range(10)`.

Reply:
I guess we also can add a `step` as a third parameter:
`Range(0,10,2)` -> 0 2 4 6 8

Reply:
That's a good idea! Let's offer all three.

Reply:
I like it!